src/person.cpp

```diff
@@ -895,6 +911,63 @@ void Person::DeActivate(AnimateResult ar)
 	this->name = nullptr;
 }
 
+
+/**
+ * Test whether this person is a guest queuing for a ride.
+ * @return Whether the person is a guest and queuing.
+ */
+bool Person::IsQueuingGuest() const
+{
+	return this->IsGuest() && static_cast<const Guest*>(this)->activity == GA_QUEUING;
+}
+
+/**
+ * Check whether another guest who is queuing for a ride is standing close to the specified position.
+ * @param vox_pos Coordinates of the voxel in the world.
+ * @param pix_pos Pixel position inside the voxel.
+ * @param only_in_front Whether to consider only guests standing in front of this one.
+ * @return Another queuing guest is close by.
+ */
+bool Person::IsQueuingGuestNearby(const XYZPoint16& vox_pos, const XYZPoint16& pix_pos, const bool only_in_front)
+{
+	/*
+	 * To ensure that guests on a neighbouring tile are also considered, we also need to check
+	 * the next voxel in all four directions, as well as the one above and the one below that.
+	 */
+	const XYZPoint32 merged_pos = MergeCoordinates(vox_pos, pix_pos);
+	std::vector<XYZPoint16> voxels = {vox_pos};
+	for (const XYZPoint16& vx : {
+			vox_pos,
+			XYZPoint16(vox_pos.x + 1, vox_pos.y, vox_pos.z),
+			XYZPoint16(vox_pos.x - 1, vox_pos.y, vox_pos.z),
+			XYZPoint16(vox_pos.x, vox_pos.y + 1, vox_pos.z),
+			XYZPoint16(vox_pos.x, vox_pos.y - 1, vox_pos.z)}) {
+		if (!IsVoxelstackInsideWorld(vx.x, vx.y)) continue;
+
+		for (const XYZPoint16& checkme : {vx, XYZPoint16(vx.x, vx.y, vx.z + 1), XYZPoint16(vx.x, vx.y, vx.z - 1)}) {
+			const Voxel *voxel = _world.GetVoxel(checkme);
+			if (voxel == nullptr) continue;
+
+			for (VoxelObject *v = voxel->voxel_objects; v != nullptr; v = v->next_object) {
+				if (v == this) continue;
```

Comment:
This looks somewhat simplistic.

Queuepaths should have only 2 edges to walk on/off the tile, so you know the relevant edges to check of the path tile. Also, it has a slope, so there is no need to brute-force try all levels as far as I can see.
Unless I have been extremely stupid, queue paths use the same system as normal paths, so there should be code to get the neighbour tile.

If the guest isn't near an edge, you can even avoid checking the neighbour tiles completely.

I am also somewhat worried about having several independent queues next to each other, but I guess that works due to "not near enough" for the neighbour check.

From an efficiency point of view, it's likely better to run this check by following the queue paths instead of doing this for each person, don't know how feasible this is.


For an initial step the current code seems sufficient though. Great work.